c_src/fast_pbkdf2.c

```diff
@@ -13,519 +14,432 @@
-{
-  write32_be(ctx->h[0], out);
-  write32_be(ctx->h[1], out + 4);
-  write32_be(ctx->h[2], out + 8);
-  write32_be(ctx->h[3], out + 12);
-  write32_be(ctx->h[4], out + 16);
-  write32_be(ctx->h[5], out + 20);
-  write32_be(ctx->h[6], out + 24);
-  write32_be(ctx->h[7], out + 28);
+static inline void write32_be(uint32_t n, uint8_t out[4]) {
+#if defined(__GNUC__) && __GNUC__ >= 4 && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+    *(uint32_t *)(out) = __builtin_bswap32(n);
+#else
+    out[0] = (n >> 24) & 0xff;
+    out[1] = (n >> 16) & 0xff;
+    out[2] = (n >> 8) & 0xff;
+    out[3] = n & 0xff;
+#endif
 }
 
-static inline void sha256_cpy(SHA256_CTX *restrict out, const SHA256_CTX *restrict in)
-{
-  out->h[0] = in->h[0];
-  out->h[1] = in->h[1];
-  out->h[2] = in->h[2];
-  out->h[3] = in->h[3];
-  out->h[4] = in->h[4];
-  out->h[5] = in->h[5];
-  out->h[6] = in->h[6];
-  out->h[7] = in->h[7];
+/* Prepare block (of blocksz bytes) to contain md padding denoting a msg-size
+ * message (in bytes).  block has a prefix of used bytes.
+ * Message length is expressed in 32 bits (so suitable for sha1, sha256, sha512). */
+static inline void md_pad(uint8_t *block, size_t blocksz, size_t used, size_t msg) {
+    memset(block + used, 0, blocksz - used - 4);
+    block[used] = 0x80;
+    block += blocksz - 4;
+    write32_be((uint32_t)(msg * 8), block);
 }
 
-static inline void sha256_xor(SHA256_CTX *restrict out, const SHA256_CTX *restrict in)
-{
-  out->h[0] ^= in->h[0];
-  out->h[1] ^= in->h[1];
-  out->h[2] ^= in->h[2];
-  out->h[3] ^= in->h[3];
-  out->h[4] ^= in->h[4];
-  out->h[5] ^= in->h[5];
-  out->h[6] ^= in->h[6];
-  out->h[7] ^= in->h[7];
+ERL_NIF_TERM mk_error(ErlNifEnv *env, const char *error_msg) {
+    return enif_make_tuple2(env, enif_make_atom(env, "error"), enif_make_atom(env, error_msg));
 }
 
-DECL_PBKDF2(sha224,
-            SHA256_CBLOCK,
-            SHA224_DIGEST_LENGTH,
-            SHA256_CTX,
-            SHA224_Init,
-            SHA224_Update,
-            SHA256_Transform,
-            SHA224_Final,
-            sha256_cpy,
-            sha224_extract,
-            sha256_xor)
-
-DECL_PBKDF2(sha256,
-            SHA256_CBLOCK,
-            SHA256_DIGEST_LENGTH,
-            SHA256_CTX,
-            SHA256_Init,
-            SHA256_Update,
-            SHA256_Transform,
-            SHA256_Final,
-            sha256_cpy,
-            sha256_extract,
-            sha256_xor)
-
-static inline void sha384_extract(SHA512_CTX *restrict ctx, uint8_t *restrict out)
-{
-  write64_be(ctx->h[0], out);
-  write64_be(ctx->h[1], out + 8);
-  write64_be(ctx->h[2], out + 16);
-  write64_be(ctx->h[3], out + 24);
-  write64_be(ctx->h[4], out + 32);
-  write64_be(ctx->h[5], out + 40);
-}
+typedef struct {
+    EVP_MD_CTX *inner;
+    EVP_MD_CTX *outer;
+} HMAC_md_ctx;
+
+#define DECL_PBKDF2(_name, _blocksz, _hashsz, _iters_per_slot)                                     \
```

Comment:
While I understand that this has changed completely, there was a useful description here. Without it, the code is harder to understand.